Source/Core/VideoCommon/HiresTextures_DDSLoader.cpp

```diff
@@ -0,0 +1,255 @@
+  level.width = header.dwWidth;
+  level.height = header.dwHeight;
+
+  // Currently, we only handle compressed textures here, and leave the rest to the SOIL loader.
+  // In the future, this could be extended, but these isn't much benefit in doing so currently.
+  // TODO: DX10 extension header handling.
+  u32 block_size = 1;
+  u32 bytes_per_block = 4;
+  bool needs_s3tc = false;
+  if (header.ddspf.dwFourCC == MAKEFOURCC('D', 'X', 'T', '1'))
+  {
+    level.format = HostTextureFormat::DXT1;
+    block_size = 4;
+    bytes_per_block = 8;
+    needs_s3tc = true;
+  }
+  else if (header.ddspf.dwFourCC == MAKEFOURCC('D', 'X', 'T', '3'))
+  {
+    level.format = HostTextureFormat::DXT3;
+    block_size = 4;
+    bytes_per_block = 16;
+    needs_s3tc = true;
+  }
+  else if (header.ddspf.dwFourCC == MAKEFOURCC('D', 'X', 'T', '5'))
+  {
+    level.format = HostTextureFormat::DXT5;
+    block_size = 4;
+    bytes_per_block = 16;
+    needs_s3tc = true;
+  }
+  else
+  {
+    // Leave all remaining formats to SOIL.
+    return false;
+  }
+
+  // We also need to ensure the backend supports these formats natively before loading them,
+  // otherwise, fallback to SOIL, which will decompress them to RGBA.
+  if (needs_s3tc && !g_ActiveConfig.backend_info.bSupportsST3CTextures)
+    return false;
+
+  // Mip levels smaller than the block size are padded to multiples of the block size.
+  u32 blocks_wide = std::max(level.width / block_size, 1u);
+  u32 blocks_high = std::max(level.height / block_size, 1u);
+
+  // Pitch can be specified in the header, otherwise we can derive it from the dimensions. For
+  // compressed formats, both DDS_HEADER_FLAGS_LINEARSIZE and DDS_HEADER_FLAGS_PITCH should be
+  // set. See https://msdn.microsoft.com/en-us/library/windows/desktop/bb943982(v=vs.85).aspx
+  if (header.dwFlags & DDS_HEADER_FLAGS_PITCH && header.dwFlags & DDS_HEADER_FLAGS_LINEARSIZE)
+  {
+    // Convert pitch (in bytes) to texels/row length.
+    if (header.dwPitchOrLinearSize < bytes_per_block)
+    {
+      // Likely a corrupted or invalid file.
+      return false;
+    }
+
+    level.row_length = std::max(header.dwPitchOrLinearSize / bytes_per_block, 1u) * block_size;
+    level.data_size = static_cast<size_t>(level.row_length / block_size) * block_size * blocks_high;
+  }
+  else
+  {
+    // Assume no padding between rows of blocks.
+    level.row_length = blocks_wide * block_size;
+    level.data_size = blocks_wide * static_cast<size_t>(bytes_per_block) * blocks_high;
+  }
+
+  // Check for truncated or corrupted files.
+  size_t data_offset = sizeof(magic) + sizeof(DDS_HEADER);
+  if ((data_offset + level.data_size) > buffer.size())
+    return false;
+
+  // Copy to the final storage location. The deallocator here is simple, nothing extra is
+  // needed, compared to the SOIL-based loader.
+  level.data = SOILPointer(new u8[level.data_size], [](u8* data) { delete[] data; });
```

Comment:
If this isn't using `SOIL`'s allocation and deallocation functions, this shouldn't be using the `SOILPointer` type or should be changed to a different name entirely.